components/brave_search/browser/brave_search_default_host.cc

```diff
@@ -0,0 +1,162 @@
+#include "brave/components/weekly_storage/daily_storage.h"
+#include "components/prefs/pref_registry_simple.h"
+#include "components/prefs/pref_service.h"
+#include "components/search_engines/template_url.h"
+#include "components/search_engines/template_url_service.h"
+
+namespace {
+
+TemplateURL* GetSearchTemplateForSite(TemplateURLService* service,
+                                      const std::string& host) {
+  // Prefer built-in entries, then offer site-provided entries,
+  // but ignore extension-provided entries.
+  TemplateURL::TemplateURLVector urls = service->GetTemplateURLs();
+  TemplateURL* other_entry = nullptr;
+  for (auto* template_url : urls) {
+    if (template_url->url_ref().GetHost(SearchTermsData()) != host) {
+      continue;
+    }
+    if (service->ShowInDefaultList(template_url)) {
+      return template_url;
+    } else if (template_url->type() != TemplateURL::OMNIBOX_API_EXTENSION &&
+               other_entry == nullptr) {
+      other_entry = template_url;
+    }
+  }
+  return other_entry;
+}
+
+}  // namespace
+
+namespace brave_search {
+
+// static
+void BraveSearchDefaultHost::RegisterProfilePrefs(
+    PrefRegistrySimple* registry) {
+  registry->RegisterListPref(brave_search::prefs::kDailyAsked);
+  registry->RegisterIntegerPref(brave_search::prefs::kTotalAsked, 0);
+}
+
+BraveSearchDefaultHost::BraveSearchDefaultHost(
+    const std::string& host,
+    TemplateURLService* template_url_service,
+    PrefService* prefs)
+    : host_(host), template_url_service_(template_url_service), prefs_(prefs) {}
+
+BraveSearchDefaultHost::~BraveSearchDefaultHost() {}
+
+bool BraveSearchDefaultHost::CanSetDefaultSearchProvider(TemplateURL* provider,
+                                                         bool is_historic) {
+  // Return false if:
+  //   - current site is default OR
+  //   - has asked more than X times in Y period OR
+  //   - search setting is controlled by extension OR
+  //   - search setting is controlled by group policy OR
+  //   - is incognito / private OR
+  // Otherwise return true
+  if (template_url_service_->IsExtensionControlledDefaultSearch()) {
+    VLOG(1) << "CanSetDefaultSearchProvider: "
+            << "Extension is controlling search engine.";
+    return false;
+  }
+  if (template_url_service_->is_default_search_managed()) {
+    VLOG(1) << "CanSetDefaultSearchProvider: "
+            << "Group Policy is controlling search engine.";
+    return false;
+  }
+  if (!provider) {
+    VLOG(1) << "CanSetDefaultSearchProvider: "
+            << "Site has not created search engine: " << host_;
+    return false;
+  }
+  const auto* default_provider =
+      template_url_service_->GetDefaultSearchProvider();
+  const bool has_default_provider = (default_provider != nullptr);
+  bool is_default =
+      (has_default_provider && default_provider->id() == provider->id());
+  if (is_default) {
+    return false;
+  }
+  if (!template_url_service_->CanMakeDefault(provider)) {
```

Comment:
Is this where the Tor/private mode check is carried out?

Reply:
@ShivanKaul no, for those profiles, `BraveSearchDefaultHost` won't be created. Instead `BraveSearchDefaultHostPrivate` will, which always returns `false`: https://github.com/brave/brave-core/pull/8819/files#diff-594631d3a40b383b831cd6c6c401ff278c4b259ec7f1b321808e868ccfb06f2bR252-R257